enhancements/olm/safe-cluster-upgrades.md

```diff
@@ -0,0 +1,270 @@
+
+## Implementation History
+
+- Initial enhancement proposal created.
+
+## Drawbacks
+
+- OpenShift users might be upset when a Cluster Upgrade fails if they are unfamiliar with OLM. These customers may ask why a Cluster Upgrade is rejected by the presence of an operator. Typically, forcing an upgrade is not fully supported and these customers may blame OLM for preventing a conventional upgrade.
+- There are instances where an Operator might not be supported on the targeted version that the operator is being upgraded to. If the customer is unable to remove the existing operator, this may place them in a position where they must choose between upgrading the cluster or removing the operator. This is arguably better than upgrading the cluster only to find out that a core service is no longer available on cluster, but I suspect that multiple tickets will be opened against OLM and the operators it installs when the customer finds themselves unable to upgrade.
+
+## Possible Future Work
+
+### Determine upgrade safety using OLM Managed Operator RBAC
+
+In addition to giving Operator Authors the ability to provide the Max OpenShift Version, OLM could determine upgrade readiness based on default OpenShift GVKs and those required by installed Operators. At a high level, the following workflow could be implemented:
+
+1. CVO provides OLM with a List of Default GVKs at OpenShift version `t`.
+2. OLM retrieves the list of default GVKs available on OpenShift version `t`. This information should be provided by the CVO team, they could possibly retrieve it from the same source that stores OpenShift upgrade channels.
+3. OLM would would generate a list of required APIs by aggregating the GVKs defined as permissions and ClusterPermissions in existing ClusterServiceVersions. Any operators that uses a wildcard (*) for the Group, Version, or Kind in its permissions/clusterPermissions would automatically get flagged as OLM would be unable to guarantee that all GVK required by the operator would be available on cluster version `t`.
+4. OLM would then compare the list of apis on the cluster at version `t` against those required by installed OLM Managed Operators. If a required GVK is unavailable, OLM would report that it is no longer upgradeable. If the required GVKs are available, OLM would mark itself as upgradeable, possibly using a special Message/Reason to signified that it has compared required GVKs against available GVKs at target version. If OLM is unable to guarantee that the required GVKs are available on OpenShift version `t` that could be highlighted as well, the cluster admin could then force the upgrade if they feel comfortable doing so.
+
+#### Pros
+
+- This feature is available by default and does not require any additional effort from Operator Authors, giving some level of gurantee that the upgrade is safe.
```

Comment:
nit: "gurantee" -> "assurance"

Reply:
Addressed in the latest version of the enhancement.